cfme/physical/physical_server.py

```diff
@@ -68,58 +69,91 @@ def load_details(self, refresh=False):
             view.browser.refresh()
             view.flush_widget_cache()
 
-    def execute_button(self, button_group, button, handle_alert=False):
+    def _execute_button(self, button, option, handle_alert=False):
         view = navigate_to(self, "Details")
-        view.toolbar.custom_button(button_group).item_select(button, handle_alert=handle_alert)
+        view.toolbar.custom_button(button).item_select(option, handle_alert=handle_alert)
+        return view
 
-    def power_on(self):
-        view = navigate_to(self, "Details")
-        view.toolbar.power.item_select("Power On", handle_alert=True)
+    def _execute_action_button(self, button, option, handle_alert=True, **kwargs):
+        target = kwargs.get('target', None)
+        provider = kwargs.get('provider', None)
+        desired_state = kwargs.get('desired_state', None)
+
+        view = self._execute_button(button, option, handle_alert=handle_alert)
+
+        if desired_state:
+            self._wait_for_state_change(desired_state, target, provider, view)
+        elif handle_alert:
+            wait_for(
+                lambda: view.flash.is_displayed,
+                message="Wait for the handle alert to appear...",
+                num_sec=5,
+                delay=2
+            )
+
+    def power_on(self, **kwargs):
+        self._execute_action_button('Power', 'Power On', **kwargs)
 
-    def power_off(self):
+    def power_off(self, **kwargs):
+        self._execute_action_button('Power', 'Power Off', **kwargs)
+
+    def power_off_immediately(self, **kwargs):
+        self._execute_action_button('Power', 'Power Off Immediately', **kwargs)
+
+    def restart(self, **kwargs):
+        self._execute_action_button('Power', 'Restart', **kwargs)
+
+    def restart_immediately(self, **kwargs):
+        self._execute_action_button('Power', 'Restart Immediately', **kwargs)
+
+    def refresh(self, provider, handle_alert=False):
+        last_refresh = provider.last_refresh_date()
+        option = 'Refresh Relationships and Power States'
+        self._execute_button('Configuration', option, handle_alert)
+        wait_for(
+            lambda: last_refresh != provider.last_refresh_date(),
+            message="Wait for the server to be refreshed...",
+            num_sec=300,
+            delay=5
+        )
+
+    def lifecycle_provision(self):
         view = navigate_to(self, "Details")
-        view.toolbar.power.item_select("Power Off", handle_alert=True)
+        view.toolbar.lifecycle.item_select("Provision Physical Server")
```

Comment:
**REQUIRED**
I don't understand why do you need this method at all? You have `Provision` navigation step and it does completely the same.

Reply:
I prefer use the method to keep a pattern in the tests, so all the tests are invoking an action through the button in the physical server page and if I use directly `navigate_to` in the test file I need to import more libraries and I break the pattern.  The Lifecycle is a button in the physical server page and when I select `Provision Physical Server`  the action is navigate to a new view, the same way occurs when I select the Power ON option inside the Power button. The action is queue a job to power on the server and if I can create a method to power_on, I think that I can create a method to select the Lifecycle button as well.

Reply:
I don't see any problem with importing `navigate_to`. Please follow DRY principle and remove this method.